### PR TITLE
Fetch docs from behaviour callback docs

### PR DESCRIPTION
**Pull request by sasa1977**
This PR aims to improve docs fetching when docs are not available in the module itself, but they can be derived from callback docs of the behaviour module.

A typical example of this is the repo module. For example, currently, vscode-elixirls produces the following:

image

With these changes, it can fetch docs from the behaviour module:

image

Testing this manually with VSCode turned out to be quite a challenge. It appears that this project moved much further compared to elixir-ls, so I ultimately had to create a temporary branch in the past. Even with that, I got some compilation warnings in elixir-ls, and some exceptions in VSCode, but this particular functionality seems to work correctly. My impression is that these exceptions are a consequence of sync loss between elixir-ls master and elixir_sense, but I'm not quite sure.

In any case, I'd advise you to test this yourself before you merge it.

### CODE REVIEW

**Comment by lukaszsamson** [comment]
Great work. I had this on my todo list. I'll need to test if it plays nicely with completions, signatures, definitions, protocols, erlang behaviours and elixir behaviours with macrocallbacks.

**Comment by lukaszsamson** [comment]
Besides macrocallback that you already fixed I noticed:

- a crash on erlang behaviour - I created a module with `@behaviour :gen_statem`
```
       ** (Protocol.UndefinedError) protocol Enumerable not implemented for nil of type Atom
            (elixir 1.10.1) lib/enum.ex:1: Enumerable.impl_for!/1
            (elixir 1.10.1) lib/enum.ex:141: Enumerable.reduce/3
            (elixir 1.10.1) lib/stream.ex:1609: Enumerable.Stream.do_each/4
            (elixir 1.10.1) lib/stream.ex:956: Stream.do_enum_transform/7
            (elixir 1.10.1) lib/stream.ex:1609: Enumerable.Stream.do_each/4
            (elixir 1.10.1) lib/enum.ex:3383: Enum.into/4
            (elixir_sense 1.0.1) lib/elixir_sense/core/normalized/code.ex:106: ElixirSense.Core.Normalized.Code.get_fun_docs/2
            (language_server 0.2.28) lib/language_server/source_file.ex:118: ElixirLS.LanguageServer.SourceFile.function_line/3
            (language_server 0.2.28) lib/language_server/dialyzer/success_typings.ex:15: anonymous fn/5 in ElixirLS.LanguageServer.Dialyzer.SuccessTypings.suggest_contracts/2
            (stdlib 3.11.2) maps.erl:232: :maps.fold_1/3
            (elixir 1.10.1) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
            (language_server 0.2.28) lib/language_server/dialyzer/success_typings.ex:10: ElixirLS.LanguageServer.Dialyzer.SuccessTypings.suggest_contracts/2
            (language_server 0.2.28) lib/language_server/dialyzer.ex:154: ElixirLS.LanguageServer.Dialyzer.handle_call/3
            (stdlib 3.11.2) gen_server.erl:661: :gen_server.try_handle_call/4
            (stdlib 3.11.2) gen_server.erl:690: :gen_server.handle_msg/6
            (stdlib 3.11.2) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```
- shows `false` instead of docs when callback has `@impl` attribute. This is related to a dubious elixir design choice - adding `@impl` automatically sets `@doc false`
- doesn't work when completions/signature are from local file metadata. Unfortunately those need separate handling.

**Comment by sasa1977** [comment]
Addressed the first two issues. I'm not really sure I understand what you mean in the third one.

**Comment by lukaszsamson** [comment]
@sasa1977 It's more like a current limitation. When listing completions or signatures for a function defined in current file we currently won't provide any docs. This happens because we use metadata generated by `ElixirSense.Core.Parser` instead of `Code.fetch_docs` and `module_info` reflection - the file edited is most likely not compiled and possibly has syntax errors. Currently `ElixirSense.Core.Parser` simply skips `@doc` attributes and completions and signatures providers return empty doc string. Let's not handle it in this PR as it's going to be a big change.

File: lib/elixir_sense/core/normalized/code.ex [hunk 1]
```diff
@@ -93,4 +90,55 @@ defmodule ElixirSense.Core.Normalized.Code do
   defp extract_docs(%{"en" => docs_en}), do: docs_en
   defp extract_docs(:hidden), do: false
   defp extract_docs(:none), do: nil
+
+  defp get_fun_docs(module, docs) do
```
Comment (lukaszsamson) [inline lib/elixir_sense/core/normalized/code.ex]: @msaraiva are you OK with introducing custom changes to this module? Before that all code in `Normalized` used to be a thin portability wrapper over stdlib stuff

Comment (sasa1977) [reply lib/elixir_sense/core/normalized/code.ex]: In case it helps, iex [does something similar](https://github.com/elixir-lang/elixir/blob/27bd9ffcc607b74ce56b547cb6ba92c9012c317c/lib/iex/lib/iex/introspection.ex#L458-L462). I presume that ex_doc also does this (but didn't test it).

Comment (msaraiva) [reply lib/elixir_sense/core/normalized/code.ex]: > code in `Normalized` used to be a thin portability wrapper over stdlib stuff

Indeed. That's was the intention. At some point, especially between v1.4 and v1.7, it was really hard to keep track of conditionals related to different versions of Elixir and then at some point, I decided to search for them in the code and move everything to `Normalized`. My dream was to someday the language would stabilize and I'd be able to remove the whole thing :)

I believe that day has come. I took a quick look at all the functions there and I think it's time to get rid of them and start using only the newer built-in counterparts. As far as I could see, we can do that if we drop support for Elixir < v1.7.

Regarding @sasa1977's changes, I think we can leave there for now and then, when we finally get rid of  `Normalized.Code`, we move them to someplace else.

Just to make it clear, I think we can still have `Normalized` for other things that might appear in the near future, however, keeping them for more than 2 minor updates makes it harder to maintain the rest of the code. Since Elixir is already on v1.10, I guess we are allowed to let it go ;)

WDYT?

Comment (sasa1977) [reply lib/elixir_sense/core/normalized/code.ex]: >  keeping them for more than 2 minor updates makes it harder to maintain the rest of the code

I'm not completely sure about the circumstances of this project, but in general my position on OSS, especially if it's maintained in people's own free time is that it's a "best effort". What this means in the context of Elixir versions is that people who want to use this lib on older Elixir versions can always lock to some past sha (having a tag provided would be nice). The new versions of the lib should be free to upgrade their lang version requirements. I think this is particularly fair in Elixir land, because upgrades to both Elixir & Erlang have in my experience always been quite painless. But even if they weren't, I'd have the same position. It's hard enough to maintain a lib for a single version of Elixir, so there have to be some strong reasons to support multiple mutually incompatible versions.

In other words, clients who want new features need to move to new Elixir (& Erlang if required). The clients who want old Elixir/Erlang compatibility won't get new features. I believe that for a software which is free of charge, this is a reasonable trade-off :-)

Comment (lukaszsamson) [reply lib/elixir_sense/core/normalized/code.ex]: > In other words, clients who want new features need to move to new Elixir (& Erlang if required). The clients who want old Elixir/Erlang compatibility won't get new features. I believe that for a software which is free of charge, this is a reasonable trade-off :-)
It's reasonable but I'd prefer to avoid python 3 mess.

**Comment by lukaszsamson** [comment]
@msaraiva I created a separate issue for tracking bump to elixir 1.7 https://github.com/elixir-lsp/elixir_sense/issues/85
Since you are OK with the changes I'm going to merge it.
